lib/buffer.js

```diff
@@ -272,30 +272,86 @@ Buffer.concat = function(list, length) {
 };
 
 
+function base64ByteLength(str, len) {
+  var bytes = len;
+
+  // Handle padding
+  if (str[len - 1] === '=')
+    bytes--;
+  if (len > 2 && str[len - 2] === '=')
+    bytes--;
+
+  // Base64 ratio: 3/4
+  bytes = (bytes / 4) * 3;
+
+  return Math.floor(bytes);
+}
+
+function utf8ByteLength(str, len) {
+  var bytes = len;
+
+  for (var i = 0; i < len; i++) {
+    var code = str.charCodeAt(i);
+
+    // Based on where the code lies (0x7F-0x10FFFF), derive char value.
+    if (code <= 0x7F)
+      { } // nop
+    else if (code <= 0x7FF)
+      bytes += 1;
+    else if (code <= 0xFFFF)
+      bytes += 2;
+    else if (code <= 0x10FFFF)
+      bytes += 3;
+  }
+
+  return bytes;
+}
+
+
 function byteLength(string, encoding) {
-  if (typeof(string) !== 'string')
-    string = String(string);
+  if (typeof string !== 'string')
+    string = '' + string;
 
-  if (string.length === 0)
+  var len = string.length;
+  if (len === 0)
     return 0;
 
   switch (encoding) {
     case 'ascii':
     case 'binary':
+    // Deprecated
     case 'raw':
-      return string.length;
+    case 'raws':
+      return len;
+
+    case 'utf8':
+    case 'utf-8':
+      return utf8ByteLength(string, len);
 
     case 'ucs2':
     case 'ucs-2':
     case 'utf16le':
     case 'utf-16le':
-      return string.length * 2;
+      return len * 2;
 
     case 'hex':
-      return string.length >>> 1;
-  }
+      return len >>> 1;
 
-  return binding.byteLength(string, encoding);
+    case 'base64':
+      return base64ByteLength(string, len);
+
+    default:
+      // The C++ binding defaulted to UTF8, we should too.
+      if (typeof encoding !== 'string')
+        return utf8ByteLength(string, len);
+
+      // Handle uppercased encodings
+      if (encoding !== encoding.toLowerCase())
+        return byteLength(string, encoding.toLowerCase());
```

Comment:
Can I suggest you cache the value of `encoding.toLowerCase()` here, to avoid calling it twice?